Common/TableProducer/caloClusterProducer.cxx

```diff
@@ -261,35 +314,417 @@ struct caloClusterProducerTask {
-                if (d < trackdist) {
-                  trackdist = d;
+                if (d < cpvdist) {
+                  cpvdist = d;
                 }
               }
             }
           }
         }
+        if (cpvdist != 99.) {      // was evaluated
+          cpvdist = sqrt(cpvdist); // was squared
+        }
+        int cpvindex = -2; // -2 no CPV in event
+        if (cpvExist) {
+          cpvindex = -1; // there were CPV clusters
+        }
+
+        float lambdaShort = 0., lambdaLong = 0.;
+        clu.getElipsAxis(lambdaShort, lambdaLong);
+
+        // Clear Collision assignment
+        if (colId == -1) {
+          // Ambiguos Collision assignment
+          cluambcursor(
+            bcMap[cluTR.getBCData().toLong()],
+            mom.X(), mom.Y(), mom.Z(), e,
+            mod, clu.getMultiplicity(), posX, posZ,
+            globaPos.X(), globaPos.Y(), globaPos.Z(),
+            clu.getTime(), clu.getNExMax(),
+            lambdaLong, lambdaShort,
+            cpvdist, cpvindex,
+            clu.firedTrigger(),
+            clu.getDistanceToBadChannel());
+
+        } else { // Normal collision
+          auto col = colls.begin() + colId;
+          clucursor(
+            col,
+            mom.X(), mom.Y(), mom.Z(), e,
+            mod, clu.getMultiplicity(), posX, posZ,
+            globaPos.X(), globaPos.Y(), globaPos.Z(),
+            clu.getTime(), clu.getNExMax(),
+            lambdaLong, lambdaShort,
+            cpvdist, cpvindex,
+            clu.firedTrigger(),
+            clu.getDistanceToBadChannel());
+        }
+      }
+    }
+  }
+
+  PROCESS_SWITCH(caloClusterProducerTask, processStandalone, "Process PHOS and CPV only", true);
+
+  //------------------------------------------------------------
+  void processFull(o2::aod::BCsWithTimestamps const& bcs,
+                   o2::aod::Collisions const& colls,
+                   o2::aod::Calos const& cells,
+                   o2::aod::CaloTriggers const& ctrs,
+                   o2::aod::CPVClusters const& cpvs,
+                   o2::aod::FullTracks const& tracks)
+  {
+
+    int64_t timestamp = bcs.begin().timestamp(); // timestamp for CCDB object retrieval
+    std::map<int64_t, int> bcMap;
+    int bcId = 0;
+    for (auto bc : bcs) {
+      bcMap[bc.globalBC()] = bcId;
+      bcId++;
+    }
+
+    // If several collisions appear in BC, choose one with largers number of contributors
+    std::map<int64_t, int> colMap;
+    int colId = 0;
+    for (auto cl : colls) {
+      auto colbc = colMap.find(cl.bc_as<aod::BCsWithTimestamps>().globalBC());
+      if (colbc == colMap.end()) { // single collision per BC
+        colMap[cl.bc_as<aod::BCsWithTimestamps>().globalBC()] = colId;
+      } else { // not unique collision per BC
+        auto coll2 = colls.begin() + colbc->second;
+        if (cl.numContrib() > coll2.numContrib()) {
+          colMap[cl.bc_as<aod::BCsWithTimestamps>().globalBC()] = colId;
+        }
+      }
+      colId++;
+    }
```

Comment:
Why chose the collision with the highest BC?
Why not sliceBy of the collision table joined event selection with o2::aod::evsel::FoundBCId to get a collision for each bc?